producer.go

```diff
@@ -106,48 +107,73 @@ type AsyncApexLogNSQHandler struct {
 // be published to.
 //
 func NewAsyncApexLogNSQHandler(marshalFunc MarshalFunc, publishFunc PublishFunc, topic string, bufferSize int) *AsyncApexLogNSQHandler {
+	var backoff = 0
 	logChan := make(chan *log.Entry, bufferSize)
 	stopChan := make(chan bool, 1)
+
+	incrementBackoff := func(backoff int) int {
+		if backoff == 0 {
+			return 1
+		}
+		newBackoff := backoff * 2
+		if newBackoff > maximumBackoffMultiple {
+			return maximumBackoffMultiple
+		}
+		return newBackoff
+	}
+
+	handler := &AsyncApexLogNSQHandler{
+		logChan:  logChan,
+		stopChan: stopChan,
+	}
+
 	go func(cLog chan *log.Entry, cStop chan bool) {
 		var e *log.Entry
 		for {
+			if backoff > 0 {
+				time.Sleep(time.Second * time.Duration(backoff))
+			}
 			select {
 			case e = <-cLog:
 				payload, err := marshalFunc(e)
 				if err != nil {
+					handler.mu.Lock()
 					backupLogger.WithError(err).Error("Marshalling log.Entry in AsyncApexLogNSQHander")
-					backupLogger.Handler.HandleLog(e)
+					backoff = incrementBackoff(backoff)
+					handler.mu.Unlock()
+
 					continue
 				}
+
 				err = publishFunc(topic, payload)
 				if err != nil {
-					backupLogger.WithError(err).Error("Publishing in AsyncApexLogNSQHander")
-					backupLogger.Handler.HandleLog(e)
+					handler.mu.Lock()
+					backupLogger.WithField("backoff", time.Second*time.Duration(backoff)).WithError(err).Error("Publishing in AsyncApexLogNSQHander")
+					backoff = incrementBackoff(backoff)
+					handler.mu.Unlock()
+
 					continue
 				}
+				backoff = 0
 			case <-cStop:
 				break
 			}
 
 		}
 	}(logChan, stopChan)
-	return &AsyncApexLogNSQHandler{
-		logChan:  logChan,
-		stopChan: stopChan,
-	}
+	return handler
 }
 
 func (h *AsyncApexLogNSQHandler) HandleLog(e *log.Entry) error {
 	h.mu.Lock()
+	defer h.mu.Unlock()
 
 	select {
 	case h.logChan <- e:
 		// Do nothing more
 	default:
```

Comment:
perhaps for another PR, but perhaps rather than using `default:` here, a `time.After(100ms or something)` would be useful here. i don't think blocking to ensure we write the log line is a bad thing.

Reply:
I'm not sure - you don't really know where you going to pay that price, or how often.